Post modin on cluster

arunjose696 opened the pull request.

arunjose696 committed:
Adding initial draft

arunjose696 committed:
adding images

arunjose696 committed:
Adding changes to appendix

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+---
+layout: post
+title: "Running Pandas on multi node cluster with Modin"
```
```suggestion
title: "Scale-out pandas beyond a single node with Modin"
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+## Introduction
+
+Handling large datasets efficiently is a common challenge in data science. Traditional tools like Pandas can struggle with scalability, often becoming slow and memory-intensive with large datasets. This is where Modin and Ray come into play. Modin is a powerful dataframe library designed to scale Pandas operations by spreading the workload across multiple cores or nodes. It supports various backe ...
```
```suggestion
Handling large datasets efficiently is a common challenge in data science.
Traditional tools like pandas can struggle with scalability, often becoming slow and
memory-intensive with large datasets. This is where Modin on Ray comes into play.
Modin is a powerful dataframe library designed to scale pandas operations by spreading the workload
across multiple cores or nodes. It supports various execution engines such as Ray, Dask and MPI,
with Ray being the default choice. Ray a distributed computing framework ensures smooth and
efficient data processing.
```

YarShev replied:
Let's use lower case for pandas. Also, please split long lines into multiple ones.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+author: arunjose696
+---
+
+## Introduction
```
```suggestion
## Annotation
```

Let's follow for previous posts.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+Handling large datasets efficiently is a common challenge in data science. Traditional tools like Pandas can struggle with scalability, often becoming slow and memory-intensive with large datasets. This is where Modin and Ray come into play. Modin is a powerful dataframe library designed to scale Pandas operations by spreading the workload across multiple cores or nodes. It supports various backe ...
+
+### Modin
+Modin is designed to scale Pandas seamlessly, allowing you to utilize all available hardware resources with minimal code changes. Simply replace your Pandas import with Modin, and you're ready to go.
```
```suggestion
### Modin

Modin is designed to scale pandas seamlessly, allowing you to utilize all available hardware resources
with minimal code changes. Simply replace your pandas import with Modin, and you're ready to go.
```

Let's put python code with import change.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+Modin is designed to scale Pandas seamlessly, allowing you to utilize all available hardware resources with minimal code changes. Simply replace your Pandas import with Modin, and you're ready to go.
+
+### Ray
+Ray is the default backend for Modin, and it shines in handling data across multiple nodes. Even when running in local mode (without a full cluster), Modin sets up a local cluster using Dask or Ray to manage the workload. One of Ray's standout features is its ability to extend across multiple nodes, allowing you to process large data workloads using a cluster of smaller machines instead of relyin ...
```
```suggestion
### Ray

Ray is the default execution engine in Modin, and it shines in handling data across multiple nodes.
Even when running locally in a single node, Modin sets up a local Ray cluster to manage the workload.
One of Ray's standout features is its ability to extend across multiple nodes, allowing you to process large
data workloads using a cluster of smaller machines instead of relying on a single, massive machine.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+### Ray
+Ray is the default backend for Modin, and it shines in handling data across multiple nodes. Even when running in local mode (without a full cluster), Modin sets up a local cluster using Dask or Ray to manage the workload. One of Ray's standout features is its ability to extend across multiple nodes, allowing you to process large data workloads using a cluster of smaller machines instead of relyin ...
+
+In contrast, Pandas operates on a single thread and node, which means that for memory-intensive tasks, it can quickly run into memory errors if a single node isn't sufficient. To illustrate this, I used the NY taxi dataset, expanding it synthetically by replicating it 10 times to create a dataset large enough to push the limits of a single-node setup.
```
```suggestion
In contrast, pandas operates on a single thread and node, which means that for memory-intensive tasks,
it can quickly run into memory errors if a single node isn't sufficient. To illustrate this, we used the NY Taxi dataset,
expanding it synthetically by replicating it 10 times to create a dataset large enough to push the limits of a single-node setup.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+
+### Work load
+For my experiments, I'll be using a script that involves three operations: `read_csv`, `apply`, and `map`. These operations have been selected to minimize data transfer between nodes, ensuring that the primary focus is on executing the workload efficiently within a cluster. The aim is to optimize performance by leveraging cluster capabilities without the added complexity of data movement between  ...
```
```suggestion
### Workload

For the experiments, we will be using a script that involves three operations: `read_csv`, `apply`, and `map`.
These operations have been selected to minimize data transfer between nodes, ensuring that the primary
focus is on executing the workload efficiently within a cluster. The aim is to optimize performance by
leveraging cluster capabilities without the added complexity of data movement between nodes.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+  </code></pre>
+</details>
+
+### Running script on pandas 
```
```suggestion
### Running script with pandas 
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+### Running script on pandas 
+
+To run this workload, I chose a t3.2xlarge instance with 32GB of RAM, enough to hold all the data in memory. However, when processing the script, Pandas encountered an Out of Memory (OOM) error and crashed because the processing dataframe required more memory than a single node could provide. The original data fit into memory, but the additional memory needed during processing caused the issue. S ...
```
```suggestion
To run this workload, we chose a t3.2xlarge instance with 32GB of RAM, enough to hold all the data in memory.
However, when processing the script, pandas encountered an Out of Memory (OOM) error and crashed
because the processing dataframe required more memory than a single node could provide.
The original data fits into memory, but the additional memory needed during processing caused the issue.
Since a single node was insufficient for the workload, we had two options: scale horizontally
(using multiple nodes) or vertically (using a more powerful machine). pandas, limited to a single node,
can't scale horizontally. Fortunately, Modin on Ray can distribute the workload across multiple nodes,
making horizontal scaling possible.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+Modin makes it a breeze to run your existing Pandas code on a cluster with minimal changes. You can simply submit your Python script as a Ray job to a Ray cluster. Just keep in mind a couple of prerequisites:
+
+1.  You need to have a Ray cluster up and running.
+2.  The dataset must be available(not just accessible) on all nodes in the cluster.
```
```suggestion
Modin makes it a breeze to run your existing pandas code in a cluster with minimal changes.
You can simply submit your Python script as a Ray job to a Ray cluster. Just keep in mind a couple of prerequisites:

1.  You need to have a Ray cluster up and running.
2.  The dataset must be available (not just accessible) on all nodes in the cluster.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+#### Setting Up Ray Cluster and Distributing the Dataset
+
+There are several ways to set up a Ray cluster. You can opt for an on-premise setup or take the easier route of configuring a cluster in AWS using tools like [KubeRay](https://github.com/ray-project/kuberay/tree/master/helm-chart/ray-cluster). In my configuration, I used [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) to download and prepare data on all node ...
```
```suggestion
There are several ways to set up a Ray cluster. You can opt for an on-premise setup or
take the easier route of configuring a cluster in AWS using tools like
[KubeRay](https://github.com/ray-project/kuberay/tree/master/helm-chart/ray-cluster).
In our configuration, we used [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
to download and prepare data on all nodes of the Ray cluster, ensuring that both prerequisites are effectively met.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
````diff
@@ -0,0 +1,152 @@
+```bash
+export RAY_ADDRESS=URL_OF_RAY_DASHBOARD
+```
+Now, make minor changes to your Python script by importing Pandas from Modin and initializing Ray.
````
```suggestion
Once the Ray cluster is up and running, set the `RAY_ADDRESS` environment variable
to the Ray dashboard URL on your client machine.

```bash
export RAY_ADDRESS=URL_OF_RAY_DASHBOARD
```

Now, make minor changes to your Python script by importing pandas from Modin and initializing Ray.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+
+
+You can view the complete Python script below.
```
```suggestion
You can view the complete Python script below.

```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+</details>
+
+
+### Results and discussion
```
```suggestion
### Results
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+With Pandas, the script took almost 400 seconds to execute, which serves as our baseline.
+
+
```
```suggestion

We can observe that by increasing the number of nodes in the cluster, Modin could handle the workload
effectively with 4 nodes or more. Significant performance improvements were evident as we scaled
from 4 to 32 nodes. Given that Modin operates efficiently with 4 nodes and both pandas and Modin
struggle with fewer than 3 nodes, it suggests the workload demands nearly 4 times the memory (32GB) of
a single t3.2xlarge instance.

To provide a fair comparison of the performance gains offered by Modin over pandas,
we decided to benchmark the script on an instance with over 120 GB of memory.
We chose an r6a.4xlarge instance, which boasts 128GiB of RAM.

With pandas, the script took almost 400 seconds to execute, which serves as our baseline.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+
+### Performance on a Single Node
+The graph reveals an interesting twist: Modin on a single node starts off slower than Pandas when using 4 workers or fewer. This could be because the overhead of distributing data and spinning up the Ray cluster doesn't pay off with so few workers. It's like trying to split chores among two people when you really need a full team. But here's where it gets exciting: crank up the CPU count (even ju ...
```
```suggestion

The graph reveals an interesting twist: Modin on a single node starts off slower than pandas
when using 4 workers or fewer. This could be because the overhead of distributing data and spinning up
the Ray cluster doesn't pay off with so few workers. It's like trying to split chores among two people
when you really need a full team. But here's where it gets exciting: crank up the CPU count
(even just to 8), and Modin shifts into high gear, outperforming pandas even on a single node.
Talk about a comeback!
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+### Performance on Scaling Nodes to 32
+
+The performance continues to soar as we add more nodes to the cluster. With a 32-node cluster, utilizing all 16 CPUs in each node, the script executes in a fraction of the time it takes on a single node. This results in a dramatic performance boost, showcasing the impressive scalability of Modin.
```
```suggestion
The performance continues to soar as we add more nodes to the cluster. With a 32-node cluster,
utilizing all 16 CPUs in each node, the script executes in a fraction of the time it takes on a single node.
This results in a dramatic performance boost, showcasing the impressive scalability of Modin.
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+All performance measurements for this post were made on an AWS r6a.4xlarge instance.
+
+- Modin version: 0.30.0
```
Why not 0.31.0? We were planning to put more operations in the script, right?

arunjose696 replied:
Just the docker image was not updated for running, I will rerun with 0.31.0 and update numbers.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,152 @@
+
+The performance continues to soar as we add more nodes to the cluster. With a 32-node cluster, utilizing all 16 CPUs in each node, the script executes in a fraction of the time it takes on a single node. This results in a dramatic performance boost, showcasing the impressive scalability of Modin.
+
+<img  src="imgs/blog_post_4/Modin_multiple_nodes.png" alt="Perf Results multinode"  style="display: block; margin-left: auto; margin-right: auto;">
```
Let's use .svg format for images so we can easily edit those in future when necessary. https://app.diagrams.net/

arunjose696 replied:
Added excel plots as svg, although the plots added currently just serve as placeholders as their measurements are taken with modin 0.31, As the  bug fixed in this [PR](https://github.com/modin-project/modin/pull/7356) would be available in modin 0.32   the numbers should be measured again before posting.

arunjose696 committed:
Apply suggestions from code review

Co-authored-by: Iaroslav Igoshev <Poolliver868@mail.ru>

arunjose696 committed:
PR comments

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+Modin is a powerful dataframe library designed to scale pandas operations by spreading the workload
+across multiple cores or nodes. It supports various execution engines such as Ray, Dask and MPI,
+with Ray being the default choice. Ray a distributed computing framework ensures smooth and
+efficient data processing.
```
```suggestion
with Ray being the default choice.
```

Either remove or move to Ray description below.

arunjose696 replied:
Removed

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+# import pandas as pd
+import modin.pandas as pd
+ 
+# your workload without any changes
```
```suggestion
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+with Ray being the default choice. Ray a distributed computing framework ensures smooth and
+efficient data processing.
+
+### Modin
```
```suggestion
### Modin

```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+In our configuration, we used [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
+to download and prepare data on all nodes of the Ray cluster, ensuring that both prerequisites are effectively met.
+
+Once the Ray cluster is up and running, set the  `RAY_ADDRESS` environment variable
```
Do we really need this?

arunjose696 replied:
I thought this made sense, as it tells the user how to make use of the remote ray cluster.

YarShev replied:
This is related to a Ray instance to connect to but you mention the Ray dashboard URL. Should we put away words about the dashboard?

arunjose696 replied:
To submit a job to a ray cluster the ray dashboard url has to be set for RAY_ADDRESS as mentioned here https://docs.ray.io/en/latest/cluster/running-applications/job-submission/quickstart.html#submitting-a-job

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
````diff
@@ -0,0 +1,202 @@
+```python
+import modin.pandas as pd
+import ray
+ray.init(address="auto", logging_level="WARNING")
+# your workload without any changes
````
```suggestion

ray.init(address="auto", logging_level="WARNING")
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
````diff
@@ -0,0 +1,202 @@
+ray.init(address="auto", logging_level="WARNING")
+# your workload without any changes
+```
+
+
````
```suggestion
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+
+  <pre><code class="language-python">
+    import pandas as pd
+    file_path = "/home/ray/data/big_yellow.csv"
```
```suggestion

    file_path = "/home/ray/data/big_yellow.csv"
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+    <pre><code class="language-python">
+    # import pandas as pd
+    import modin.pandas as pd
+    import ray
+    from modin.utils import execute # execute is just used to make sure all the asynchronous operations are complete as we benchmark.
```
```suggestion
    from modin.utils import execute # execute is just used to make sure all the asynchronous operations are complete as we benchmark.
        import ray

```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+</details>
+
+### Results
+
```
```suggestion
```

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,202 @@
+
+We can observe that by increasing the number of nodes in the cluster, Modin could handle the workload
+effectively with 4 nodes or more. Significant performance improvements were evident as we scaled
+from 4 to 32 nodes. Given that Modin operates efficiently with 4 nodes and both pandas and Modin
```
Is this correct to talk about pandas as to number of nodes?

arunjose696 replied:
would it be okay if we mention in brackets (as pandas does not offer horizontal scaling)?

YarShev replied:
Yes, let's put this clarification.

arunjose696 replied:
Added clarification

Retribution98 commented:
Good job. I think this is a great article for a beginner user! Can we add links to Modin tutorials?

We use a workload for perfect parallelization, and demonstrate excellent scalability. This is good, but along with that we suggest the user to use his workload. However, when he runs it, he will see completely different results and will be disappointed.

Let's add more explanations on possible problems on non-sentientic workloads.

arunjose696 committed:
Apply suggestions from code review

Co-authored-by: Iaroslav Igoshev <Poolliver868@mail.ru>

arunjose696 committed:
Update docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md

Co-authored-by: Iaroslav Igoshev <Poolliver868@mail.ru>

arunjose696 committed:
PR comments

arunjose696 committed:
adding svg images from modin 0.31 as place holders

arunjose696 commented:
> Good job. I think this is a great article for a beginner user! Can we add links to Modin tutorials?
> 
> We use a workload for perfect parallelization, and demonstrate excellent scalability. This is good, but along with that we suggest the user to use his workload. However, when he runs it, he will see completely different results and will be disappointed.
> 
> Let's add more explanations on possible problems on non-sentientic workloads.

Added a disclaimer suggesting the possiblity of results not being optimal for every workload and added a section called futher reading with link to modin tutorials

arunjose696 commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,213 @@
+
+All performance measurements for this post were made on an AWS r6a.4xlarge instance.
+
+- Modin version: 0.32.0
```
The numbers currently in the plots have no t been measured with modin 0.32.0 the numbers would be updated on release

arunjose696 replied:
Have updated the  plots with measurements from modin 0.32.0.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,213 @@
+
+### Modin
+
+Modin is designed to scale pandas seamlessly, allowing you to utilize all available hardware resources with minimal code changes. Simply replace your pandas import with Modin, and you're ready to go.
```
```suggestion
Modin is designed to scale pandas seamlessly, allowing you to utilize all available hardware resources
with minimal code changes. Simply replace your pandas import with Modin, and you're ready to go.
```

arunjose696 committed:
Update docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md

Co-authored-by: Iaroslav Igoshev <Poolliver868@mail.ru>

arunjose696 committed:
updating the plots

arunjose696 commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,214 @@
+    # import pandas as pd
+    import modin.pandas as pd
+    from modin.utils import execute # execute is just used to make sure all the asynchronous operations are complete as we benchmark.
+        import ray
+
+    ray.init(address="auto", logging_level="WARNING")
```
```suggestion
    import ray
    ray.init(address="auto", logging_level="WARNING")
```

arunjose696 commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,214 @@
+
+We can observe that by increasing the number of nodes in the cluster, Modin could handle the workload
+effectively with 4 nodes or more. Significant performance improvements were evident as we scaled
+from 4 to 32 nodes. Given that Modin operates efficiently with 4 nodes and both pandas(as pandas does not offer horizontal scaling) and Modin
```
```suggestion
from 4 to 32 nodes. Given that Modin operates efficiently with 4 nodes and both pandas (as pandas does not offer horizontal scaling) and Modin
```

arunjose696 commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
```diff
@@ -0,0 +1,214 @@
+    <pre><code class="language-python">
+    # import pandas as pd
+    import modin.pandas as pd
+    from modin.utils import execute # execute is just used to make sure all the asynchronous operations are complete as we benchmark.
```
```suggestion
    # execute is just used to make sure all the asynchronous operations are complete as we benchmark.
    from modin.utils import execute 
```

arunjose696 committed:
Apply suggestions from code review

arunjose696 commented:
@YarShev all comments addressed can we publish this.

YarShev commented on docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md:
````diff
@@ -0,0 +1,214 @@
+    execute(result)
+    </code></pre>
+
+  ```
````
```suggestion
```

YarShev committed:
Update docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md
